build-logic/cleanup/src/main/java/gradlebuild/cleanup/KillLeakingJavaProcesses.java

```diff
@@ -0,0 +1,248 @@
+        /**
+         * The pid of the process.
+         */
+        final String pid;
+        /**
+         * The command line arguments of the process. The first argument is java executable.
+         */
+        final List<String> commands = new ArrayList<>();
+        /**
+         * The main class of the process. May be null.
+         */
+        String mainClass;
+        /**
+         * The classpath passed to the process via `-cp` or `-classpath`. May be null.
+         */
+        String classpath;
+
+        private JavaProcessInfo(String pid, String javaCommand, String otherCommands) {
+            this.pid = pid;
+            commands.add(javaCommand);
+            Stream.of(otherCommands.split("\\s"))
+                .map(String::trim)
+                .filter(s -> !s.isEmpty())
+                .forEach(commands::add);
+            for (int i = 0; i < commands.size(); ++i) {
+                String command = commands.get(i);
+                if (MAIN_CLASS_PATTERN.matcher(command).matches()) {
+                    mainClass = command;
+                }
+                if (i != commands.size() - 1 && "-classpath".equals(command) || "-cp".equals(command)) {
+                    classpath = commands.get(i + 1);
+                    if (classpath.startsWith("\"")) {
+                        classpath = classpath.substring(1, classpath.length());
+                    }
+                    if (classpath.endsWith("\"")) {
+                        classpath = classpath.substring(0, classpath.length() - 1);
+                    }
+                }
+            }
+        }
+
+        static List<JavaProcessInfo> parsePsOutput(String psOutput, boolean isWindows) {
+            return psOutput.lines()
+                .map(line -> JavaProcessInfo.parseLine(line, isWindows))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+        }
+
+        /**
+         * Parse a jps output line and get the Java process info.
+         *
+         * @return null if the line doesn't match the pattern.
+         */
+        private static JavaProcessInfo parseLine(String line, boolean isWindows) {
+            Pattern pattern = isWindows ? WINDOWS_JAVA_PROCESS_PATTERN : UNIX_JAVA_PROCESS_PATTERN;
+            Matcher matcher = pattern.matcher(line.trim());
+            if (matcher.find()) {
+                if (isWindows) {
+                    return new JavaProcessInfo(matcher.group(4), matcher.group(1), matcher.group(3));
+                } else {
+                    return new JavaProcessInfo(matcher.group(1), matcher.group(2), matcher.group(3));
+                }
+            } else {
+                return null;
+            }
+        }
+
+        private boolean isLeakingJavaProcess() {
+            if (mainClass == null) {
+                return false;
+            }
+            if (mainClass.endsWith(".GradleWorkerMain") || mainClass.endsWith(".KotlinCompileDaemon")) {
```

Comment:
The `play...Netty` server is missing here.

Reply:
No longer an issue.